src/celengine/astrodb.cpp

```diff
@@ -0,0 +1,396 @@
+        return AstroCatalog::InvalidIndex;
+    return it->second->get(nr);
+}
+
+AstroCatalog::IndexNumber AstroDatabase::nameToIndex(const std::string& name, bool tryGreek, bool smart) const
+{
+    AstroCatalog::IndexNumber nr = m_nameDB.getIndexNumberByName(name, tryGreek);
+    if (nr != AstroCatalog::InvalidIndex)
+        return nr;
+    if (smart)
+    {
+        nr = m_nameDB.findIndexNumberByName(name, tryGreek);
+        if (nr != AstroCatalog::InvalidIndex)
+            return nr;
+    }
+    for (const auto& ci : m_catalogs)
+    {
+        AstroCatalog::IndexNumber inr = ci.second->nameToCatalogNumber(name);
+        if (inr == AstroCatalog::InvalidIndex)
+            continue;
+        if (ci.first == Hipparcos && inr < HipparcosAstroCatalog::MaxCatalogNumber)
+            return inr;
+        if (ci.first == Tycho && inr > HipparcosAstroCatalog::MaxCatalogNumber && inr <= TychoAstroCatalog::MaxCatalogNumber)
+            return inr;
+        nr = catalogNumberToIndex(ci.first, inr);
+    }
+    return nr;
+}
+
+AstroCatalog::IndexNumber AstroDatabase::starnameToIndex(const std::string& name, bool tryGreek) const
+{
+    return m_nameDB.findIndexNumberByName(name, tryGreek);
+}
+
+std::string AstroDatabase::catalogNumberToString(AstroCatalog::IndexNumber nr) const
+{
+    return fmt::sprintf("#%u", nr);
+}
+
+std::string AstroDatabase::catalogNumberToString(int catalog, AstroCatalog::IndexNumber nr) const
+{
+    auto it = m_catalogs.find(catalog);
+    if (it != m_catalogs.end())
+        return it->second->catalogNumberToName(nr);
+    return "";
+}
+
+std::string AstroDatabase::getObjectName(AstroCatalog::IndexNumber nr, bool i18n) const
+{
+    NameDatabase::NumberIndex::const_iterator iter = m_nameDB.getFirstNameIter(nr);
+    if (iter != m_nameDB.getFinalNameIter() && iter->first == nr)
+    {
+        if (i18n)
+        {
+            const char* localized = _(iter->second.c_str());
+            if (iter->second != localized)
+                return localized;
+        }
+        return iter->second;
+    }
+    return catalogNumberToString(nr);
+}
+
+std::vector<std::string> AstroDatabase::getObjectNameList(AstroCatalog::IndexNumber nr, int max) const
+{
+    std::vector<std::string> ret;
+    NameDatabase::NumberIndex::const_iterator iter = m_nameDB.getFirstNameIter(nr);
+    while (iter != m_nameDB.getFinalNameIter() && iter->first == nr && max > 0)
+    {
+        ret.emplace_back(iter->second);
+        ++iter;
+        --max;
+    }
+
+    if (max == 0)
+        return ret;
+    for (const auto it : m_catalogs)
+    {
+        if (max == 0)
+            break;
+        AstroCatalog::IndexNumber inr = indexToCatalogNumber(it.first, nr);
+        if (inr == AstroCatalog::InvalidIndex)
+        {
+//             cout << "Invalid cross index entry for catalog " << it.first << "[" << nr << "]\n";
+            continue;
+        }
+        ret.emplace_back(it.second->catalogNumberToName(inr));
```

Comment:
Again, it `catalogNumberToName` returns a new string, then `emplace_back` is ok and preferred, but if it return an existing object, then use `push_back`